pkg/workflows/management/upgrade_management_components.go

```diff
@@ -0,0 +1,163 @@
+	}
+}
+
+func (u *UMCValidator) PreflightValidations(ctx context.Context) []validations.Validation {
+	return []validations.Validation{
+		func() *validations.ValidationResult {
+			return &validations.ValidationResult{
+				Name:        "control plane ready",
+				Remediation: fmt.Sprintf("ensure control plane nodes and pods for cluster %s are Ready", u.cluster.Name),
+				Err:         u.kubectl.ValidateControlPlaneNodes(ctx, u.cluster, u.cluster.Name),
+			}
+		},
+		func() *validations.ValidationResult {
+			return &validations.ValidationResult{
+				Name:        "cluster CRDs ready",
+				Remediation: "",
+				Err:         u.kubectl.ValidateClustersCRD(ctx, u.cluster),
+			}
+		},
+	}
+}
+
+// Run Upgrade implements upgrade functionality for management cluster's upgrade operation.
+func (umc *UpgradeManagementComponentsRunner) Run(ctx context.Context, clusterSpec *cluster.Spec, managementCluster *types.Cluster, validator interfaces.Validator) error {
+	commandContext := &task.CommandContext{
+		Provider:          umc.provider,
+		ClusterManager:    umc.clusterManager,
+		ManagementCluster: managementCluster,
+		ClusterSpec:       clusterSpec,
+		Validations:       validator,
+		Writer:            umc.writer,
+		CAPIManager:       umc.capiManager,
+		UpgradeChangeDiff: types.NewChangeDiff(),
+		GitOpsManager:     umc.gitOpsManager,
+		EksdUpgrader:      umc.eksdUpgrader,
+		EksdInstaller:     umc.eksdInstaller,
+	}
+
+	return task.NewTaskRunner(&setupAndValidateMC{}, umc.writer).RunTask(ctx, commandContext)
+}
+
+type setupAndValidateMC struct{}
+
+// Run setupAndValidate validates management cluster before upgrade process starts.
+func (s *setupAndValidateMC) Run(ctx context.Context, commandContext *task.CommandContext) task.Task {
+	logger.Info("Performing setup and validations")
+	currentSpec, err := commandContext.ClusterManager.GetCurrentClusterSpec(ctx, commandContext.ManagementCluster, commandContext.ClusterSpec.Cluster.Name)
+	if err != nil {
+		commandContext.SetError(err)
+		return nil
+	}
+	commandContext.CurrentClusterSpec = currentSpec
+	runner := validations.NewRunner()
+	runner.Register(commandContext.Validations.PreflightValidations(ctx)...)
+
+	err = runner.Run()
+	if err != nil {
+		commandContext.SetError(err)
+		return nil
+	}
+
+	return &upgradeCoreComponentsMC{}
+}
+
+func (s *setupAndValidateMC) Name() string {
+	return "validate"
+}
+
+func (s *setupAndValidateMC) Restore(_ context.Context, _ *task.CommandContext, _ *task.CompletedTask) (task.Task, error) {
+	return nil, nil
+}
+
+func (s *setupAndValidateMC) Checkpoint() *task.CompletedTask {
+	return &task.CompletedTask{
+		Checkpoint: nil,
+	}
+}
+
+type upgradeCoreComponentsMC struct {
+	upgradeCoreComponents
+}
+
+func (u *upgradeCoreComponentsMC) Run(ctx context.Context, commandContext *task.CommandContext) task.Task {
+	ntask := u.upgradeCoreComponents.Run(ctx, commandContext)
+	if _, ok := ntask.(*workflows.CollectMgmtClusterDiagnosticsTask); !ok {
```

Comment:
I would recommend to not trying to reuse tasks across workflows. This package wasn't designed to do so. Instead, just extract the common functionality to a method and call the method from two different tasks that just act as a wrapper.

If you want to reuse tasks, look into `pkg/workflow` which was better designed to do so (and it's better in general).